libraries/Microsoft.Bot.Builder/BotFrameworkAdapter.cs

```diff
@@ -160,59 +158,60 @@ public new BotFrameworkAdapter Use(IMiddleware middleware)
-                appCredentials = new MicrosoftAppCredentials(appId, appPassword);
-                _appCredentialMap[appId] = appCredentials;
+                throw new NotSupportedException("ClaimsIdemtity cannot be null. Pass Anonymous ClaimsIdentity if authentication is turned off.");
             }
 
-            return appCredentials;
-        }
-
-        /// <summary>
-        /// Gets the bot identifier from claims.
-        /// </summary>
-        /// <param name="claimsIdentity">The claims identity.</param>
-        /// <returns>Bot's AAD AppId, if it could be inferred from claims. Null otherwise.</returns>
-        private static string GetBotId(ClaimsIdentity claimsIdentity)
-        {
-            // For Activities coming from channels Audience Claim contains the Bot's AAD AppId
-            Claim botAppIdClaim = (claimsIdentity.Claims?.SingleOrDefault(claim => claim.Type == AuthenticationConstants.AudienceClaim) 
+            // For requests from channel App Id is in Audience claim of JWT token. For emulator it is in AppId claim. For 
+            // unauthenticated requests we have anonymouse identity provided auth is disabled.
+            Claim botAppIdClaim = (claimsIdentity.Claims?.SingleOrDefault(claim => claim.Type == AuthenticationConstants.AudienceClaim)
                 ??
                 // For Activities coming from Emulator AppId claim contains the Bot's AAD AppId.
                 claimsIdentity.Claims?.SingleOrDefault(claim => claim.Type == AuthenticationConstants.AppIdClaim));
 
             // For anonymous requests (requests with no header) appId is not set in claims.
             if (botAppIdClaim != null)
             {
-                return botAppIdClaim.Value;
+                string botId = botAppIdClaim.Value;
+                MicrosoftAppCredentials appCredentials = await this.GetAppCredentialsAsync(botId);
+                return new ConnectorClient(new Uri(serviceUrl), appCredentials);
             }
             else
             {
-                return null;
+                return new ConnectorClient(new Uri(serviceUrl));
             }
         }
 
-        public static void AssertBotFrameworkContext(ITurnContext context)
+        /// <summary>
+        /// Gets the application credentials. App Credentials are cached so as to ensure we are not refreshing
+        /// token everytime.
+        /// </summary>
+        /// <param name="appId">The application identifier (AAD Id for the bot).</param>
+        /// <returns>App credentials.</returns>
+        private async Task<MicrosoftAppCredentials> GetAppCredentialsAsync(string appId)
         {
-            BotAssert.ContextNotNull(context); 
+            if (appId == null)
+            {
+                return MicrosoftAppCredentials.Empty;
+            }
+
+            if (!_appCredentialMap.TryGetValue(appId, out var appCredentials))
+            {
+                string appPassword = await _credentialProvider.GetAppPasswordAsync(appId);
```

Comment:
We should be consistent in our use of `var` vs explicit types. Right now the codebase follows a `var`-only approach so, religious debates aside, I think we should stick with that.

Reply:
wildo.